### PR TITLE
[autoscaler] Expose autoscaler container security context.

### PR DESCRIPTION
**Pull request by DmitriGekhtman**
Signed-off-by: Dmitri Gekhtman <dmitri.m.gekhtman@gmail.com>

## Why are these changes needed?

Allows setting the autoscaler container's security context.

## Checks

- [ ] I've made sure the tests are passing. 
- Testing Strategy
   - [ ] Unit tests
   - [ ] Manual tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by DmitriGekhtman** [comment]
cc @juliusvonkohout

File: ray-operator/controllers/ray/common/pod_test.go [hunk 1]
```diff
@@ -523,6 +523,22 @@ func TestBuildPodWithAutoscalerOptions(t *testing.T) {
...
 
+	// Define a custom security profile.
+	allowPrivilegeEscalation := false
+	capabilities := v1.Capabilities{
+		Drop: []v1.Capability{"All"},
+	}
+	seccompProfile := v1.SeccompProfile{
+		Type: v1.SeccompProfileTypeRuntimeDefault,
+	}
+	runAsNonRoot := false
```
Comment (kevin85421) [inline ray-operator/controllers/ray/common/pod_test.go:534]: false -> true
```
securityContext:
  allowPrivilegeEscalation: false
  capabilities:
    drop: ["ALL"]
  runAsNonRoot: true
  seccompProfile:
    type: RuntimeDefault
```

File: ray-operator/controllers/ray/common/pod_test.go [hunk 2]
```diff
@@ -523,6 +523,22 @@ func TestBuildPodWithAutoscalerOptions(t *testing.T) {
 	customEnv := []v1.EnvVar{{Name: "fooEnv", Value: "fooValue"}}
 	customEnvFrom := []v1.EnvFromSource{{Prefix: "Pre"}}
 
+	// Define a custom security profile.
+	allowPrivilegeEscalation := false
+	capabilities := v1.Capabilities{
+		Drop: []v1.Capability{"All"},
```
Comment (kevin85421) [inline ray-operator/controllers/ray/common/pod_test.go:529]: I am not sure whether "All" is equivalent with "ALL". In my experience, `drop: ["ALL"]` works, but `drop: ["all"]` does not work.

**Review by kevin85421 — COMMENTED** [review]
Thank @DmitriGekhtman for this contribution! We should create a namespace with `restricted` policies to test it. Follow the instructions in https://github.com/ray-project/kuberay/pull/750/files.

File: ray-operator/controllers/ray/common/pod_test.go (hunk 2 above)
Comment (DmitriGekhtman) [reply ray-operator/controllers/ray/common/pod_test.go]: Fixed.

File: ray-operator/controllers/ray/common/pod_test.go (hunk 1 above)
Comment (DmitriGekhtman) [reply ray-operator/controllers/ray/common/pod_test.go]: Fixed.

**Comment by DmitriGekhtman** [comment]
TODO: Try it in a namespace with restricted policies.

**Comment by DmitriGekhtman** [comment]
@kevin85421 I just tested this with your securityContext instructions and it worked exactly as expected -- I was able to deploy a cluster and trigger autoscaling.

File: ray-operator/controllers/ray/common/pod.go [hunk 3]
```diff
@@ -446,6 +446,9 @@ func mergeAutoscalerOverrides(autoscalerContainer *v1.Container, autoscalerOptio
 		if len(autoscalerOptions.EnvFrom) > 0 {
 			autoscalerContainer.EnvFrom = append(autoscalerContainer.EnvFrom, autoscalerOptions.EnvFrom...)
 		}
+		if autoscalerOptions.SecurityContext != nil {
+			autoscalerContainer.SecurityContext = autoscalerOptions.SecurityContext
```
Comment (kevin85421) [inline ray-operator/controllers/ray/common/pod.go:450]: 1. Use deep copy?
2. Do we need to check `nil`?  Maybe we just need `autoscalerContainer.SecurityContext = autoscalerOptions.SecurityContext`.

Comment (DmitriGekhtman) [reply ray-operator/controllers/ray/common/pod.go:450]: 1. Sure, let me try a deep copy.
2. Strictly speaking, you're right -- the if condition is redundant. However, I want to keep it consistent with the rest of the code. The "if" expresses "If the user has set a custom securityContext, use it to configure the autoscaler container."

Comment (DmitriGekhtman) [reply ray-operator/controllers/ray/common/pod.go:450]: Plus, I think I might change the default to the configuration proposed by you and Julius.

Comment (kevin85421) [reply ray-operator/controllers/ray/common/pod.go:450]: > Strictly speaking, you're right -- the if condition is redundant. However, I want to keep it consistent with the rest of the code. The "if" expresses "If the user has set a custom securityContext, use it to configure the autoscaler container."

Makes sense to me.

> Plus, I think I might change the default to the configuration proposed by you and Julius.

What's the meaning of "default" here? Thanks!

Comment (DmitriGekhtman) [reply ray-operator/controllers/ray/common/pod.go:450]: Actually, never mind that.

Comment (DmitriGekhtman) [reply ray-operator/controllers/ray/common/pod.go:450]: > What's the meaning of "default" here? Thanks!

Never mind!

Comment (DmitriGekhtman) [reply ray-operator/controllers/ray/common/pod.go]: Added the DeepCopy!

**Review by kevin85421 — APPROVED** [review]
LGTM

**Pull request merged**